### ISSUE TITLE
Labels omit any character outside of the Basic Multilingual Plane

### ISSUE THREAD

**Issue by 1ec5** [maplibre/maplibre-gl-js#2307]
Symbol layers can only render Unicode characters in the Basic Multilingual Plane (BMP, up to U+FFFF), but they do not render any character in one of the other non-BMP [planes](https://en.wikipedia.org/wiki/Plane_(Unicode)) (“astral planes”). GL JS simply skips the character without even leaving a space.

## Example

According to OpenStreetMap, [this tourist attraction in Taiwan](https://www.openstreetmap.org/node/4322299689) is named “卡司𥰆拉樂園” in Traditional Chinese. But [this stripped-down GL JS rendering of that label](https://jsbin.com/foroneqayo/1/edit?js,output) appears as “卡司拉樂園”, skipping over the “𥰆” (U+25C06). This is the case whether ideographic glyphs are being rendered on the server side or locally on the client side.

## Impact

As of 2017, this issue affects a variety of Chinese (particularly Cantonese) and Japanese names in OpenStreetMap: https://github.com/mapbox/mapbox-gl-js/issues/4001#issuecomment-312393003. In Chinese, skipping a character could easily confuse or offend the user. OSM also contains names in some minority languages that require astral-plane characters, such as Hmong and Osage. Some GL JS–based maps, such as OpenStreetMap Americana, allow the user to switch to a minority language globally, even where it lacks official status.

Apart from OSM, developers may want to label geographic features using historic languages and writing systems. For example, OpenHistoricalMap, which currently uses Mapbox GL JS, contains historic place names in [Egyptian hieroglyphics](https://taginfo.openhistoricalmap.org/keys/name%3Aegy) and [Sumerian cuneiform](https://taginfo.openhistoricalmap.org/keys/name%3Asux). These places are invisible, making the map much less useful in some time periods. Less seriously but perhaps more practically, a developer could choose to spice up their map with hard-coded emoji in symbol layers. Many emoji are located outside the BMP, although there are other obstacles to emoji support as well: https://github.com/mapbox/mapbox-gl-native/issues/15666#issuecomment-532969076.

## Diagnosis

The most naïve fix for this issue would entail removing some places where glyph IDs are capped at 65,535 and increasing the data type to something wider than `UInt16`:

https://github.com/maplibre/maplibre-gl-js/blob/6d55d5ba5c0b60d074ba19e80714551498c2af68/src/data/bucket/symbol_bucket.ts#L957-L963 https://github.com/maplibre/maplibre-gl-js/blob/66e1009d30c24f83b280faad53dac3aed7cc5b8b/src/render/glyph_manager.ts#L93-L96

However, there are other considerations for supporting astral-plane characters. For one thing, the Mapbox Font API or any imitator only provides glyph PBFs for glyph ranges up to U+FFFF. Implementers would need to generate more glyph PBFs for the newly supported ranges to avoid HTTP errors.

A trickier issue is that UTF-8-encoded text uses surrogate pairs (two undisplayable control characters) to represent anything outside the BMP, but GL JS assumes that a single UTF-8 character is equivalent to a single glyph ID. Wherever the code is currently converting UTF-8 text into glyph IDs, it would need to coalesce any surrogate pairs. There could be latent bugs because some common JavaScript `String` methods fail to correctly resolve surrogate pairs. For example, in some cases, the code would need to use [`codePointAt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt) instead of subscripting, `at`, or `charAt`.

The character=glyph assumption is also relevant to complex text rendering (maplibre/maplibre#231), but fortunately, supporting astral-plane characters is much more straightforward than complex text rendering, because there aren’t inherently any special text layout considerations. That said, a number of additional Unicode blocks would need to be enabled for vertical text layout and ideographic line breaking; they were omitted previously because of this missing character support:

https://github.com/maplibre/maplibre-gl-js/blob/e78ad7944ef768e67416daa4af86b0464bd0f617/src/util/script_detection.ts#L72-L73 https://github.com/maplibre/maplibre-gl-js/blob/be86cc3dcda3074ee598429153a503ab94849d49/src/util/is_char_in_unicode_block.ts#L297-L302

**Comment by HarelM** [comment]
While I agree this is important and needed, is there a workaround that can be used until this is fixed?
For example something like using special fonts that have characters in the BMP, generate pbf for them and have them on the server?
I use this method to generate icons for my buttons in HTML. this is how font awesome is built and icomoon allows easy generation of these fonts.

**Comment by 1ec5** [comment]
> For example something like using special fonts that have characters in the BMP, generate pbf for them and have them on the server?

This would be an option for custom data provided by the developer that requires a small set of known characters. However, for basemap data, it would be impractical to special-case each of the real-world features that have names requiring unsupported characters. It would also come with some downsides, such as breaking the mapbox-gl-accessibility plugin’s screenreader support.

**Comment by wipfli** [comment]
Interesting, thanks for sharing. So basically your idea @1ec5 would be to replace the 1 character = 1 code point assumption? If this is what you are suggesting, where would one need to change the code base? I guess here in GL JS, but also in native and maybe in the font api?

**Comment by 1ec5** [comment]
We need to unwind this assumption anyways for complex text rendering, whichever approach we take for that issue. Fortunately, it’s possible to solve this issue with a half-measure: keep assuming that codepoints correspond to glyph IDs, _except_ when we encounter a low surrogate followed by a high surrogate. In practice, this just means using `codePointAt` and adopting a wider data type for StructArray indices.

gl-native is affected by the same issue and a similar fix would be necessary there. However, it should be possible to make progress on one platform before another. The fix wouldn’t affect the public interface or style specification. The only externally observable effects would be HTTP requests for these higher ranges and visually correct output.

**Comment by HarelM** [comment]
Bounty direction:
https://github.com/maplibre/maplibre/issues/193
@1ec5 let me know if you would like to work on that and what bounty size you think it's appropriate.

**Comment by 1ec5** [comment]
Without additional layout changes, local SDF rendering would still need to be limited to fixed-pitch scripts such as CJK (in the SIP) and Khitan small script (in the SMP).

**Comment by bdon** [comment]
We have the logic for that already in https://github.com/maplibre/maplibre-gl-js/blob/main/src/util/is_char_in_unicode_block.ts, right?

**Comment by 1ec5** [comment]
Yes, given a reliable list of the scripts that are fixed-pitch. That would include many of the character ranges capable of [ideographic breaking or vertical writing mode](https://github.com/maplibre/maplibre-gl-js/blob/cc9d8e8229769cb35ad5761ad3a4c7f9fccf381d/src/util/script_detection.ts#L5-L16), but not all, and there are probably some fixed-width blocks besides. Anyways, my point is that this will mostly be of benefit to CJKV – not a bad thing by any means – and maybe some “emoji-adjacent” characters. Other non-BMP scripts would be quite messy even with this baby step.
